Generic arg disambiguation

yodaldevoid opened the pull request:
Using the tactic suggested by @petrochenkov in https://github.com/rust-lang/rust/issues/60804#issuecomment-516769465 and on [zulip](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/generic.20argument.20disambiguation), this change checks type arguments to see if they are really incorrectly-parsed const arguments.

it should be noted that `segments.len() == 1 && segments[0].arg.is_none()` was reduced to `segments.len() == 1` as suggested by @petrochenkov in [zulip](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/generic.20argument.20disambiguation/near/177848002). This change allowed a few more existing tests to have their braces removed.

There are a couple of "problems" with these changes that I should note. First, there was a regression in the error messages found in "src/test/ui/privacy-ns1.rs" and "src/test/ui/privacy-ns1.rs". Second, some braces were unable to be removed from "src/test/ui/const-generics/fn-const-param-infer.rs". Those on line 24 caused the statement to stop equating when removed, and those on line 20 cause a statement that should not equate to produce no error when removed.

I have not looked further into any of these issues yet, though I would be willing to look into them before landing this. I simply wanted to get some other eyes on this before going further.

Fixes #60804

cc @varkor @jplatte

rust-highfive commented:
r? @eddyb

(rust_highfive has picked a reviewer for you, use r? to override)

jplatte commented on src/librustc/hir/lowering.rs:
```diff
@@ -1149,13 +1149,80 @@ impl<'a> LoweringContext<'a> {
+                // we resolved the path in the value namespace, we transform it into a generic
+                // const argument.
+                if let TyKind::Path(ref qself, ref path) = ty.kind {
+                    // FIXME: the `is_none` may be redundant. Try without it.
```
Outdated comment

jplatte commented on src/librustc_resolve/late.rs:
```diff
@@ -544,6 +544,51 @@ impl<'a, 'tcx> Visitor<'tcx> for LateResolutionVisitor<'a, '_> {
+                            ty.span,
+                            CrateLint::SimplePath(ty.id),
+                        ) {
+                            // FIXME: maybe use `resolve_qpath`
```
Another outdated comment

eddyb commented on src/librustc/hir/lowering.rs:
```diff
@@ -1149,13 +1149,79 @@ impl<'a> LoweringContext<'a> {
+                        if let Some(partial_res) = self.resolver.get_partial_res(ty.id) {
+                            match partial_res.base_res() {
+                                // FIXME(const-generics): Currently matches against all items in
+                                // the value namespace. Maybe this should be pared down?
```
I wonder if we could automate getting the namespace so we don't have to match on the cases here.

yodaldevoid replied:
I looked at implementing a `is_valid_in_ns` function on `Res`, but didn't go with that implementation because I didn't want to leave out `Res::Err` or `Res::Def(DefKind::Ctor(_, CtorKind::Fictive)`. Both are valid in the value namespace (to my knowledge), but not here.

Thinking on it, I suppose I could have specifically checked for those two variants and still implemented the function... Don't code at 3 a.m., kids.

eddyb commented on src/test/ui/const-generics/const-generic-array-wrapper.rs:
```diff
@@ -3,11 +3,11 @@
+struct Foo<T, const N: usize>([T; N]);
 
-impl<T, const N: usize> Foo<T, {N}> {
+impl<T, const N: usize> Foo<T, N> {
```
This is the only use in this file that this PR allows, I'm surprised the others were written `{N}` (as they never needed to be)

eddyb commented on src/test/ui/const-generics/fn-const-param-call.rs:
```diff
@@ -9,12 +9,12 @@ fn function() -> u32 {
 
 fn main() {
-    assert_eq!(Wrapper::<{function}>::call(), 17);
+    assert_eq!(Wrapper::<function>::call(), 17);
```
Oh wow this is really cool (even if I'm not sure it will survive `structural_match` requirements).

yodaldevoid replied:
Yeah, I am not worrying too much about this test as I also have a feeling most of this won't survive putting in place the `structural_match` requirement. As a side note, I was looking at implementing that requirement soon&#8482;.

varkor replied:
That was implemented in https://github.com/rust-lang/rust/pull/6562. Function types count as `structural_match`, though, as far as I can tell.

yodaldevoid replied:
@varkor I'm not seeing a reference to a PR in that issue. Can you point to the PR that implemented the `structural_match` constraint?

varkor replied:
(Sorry, my phone keyboard likes deleting the last character of pasted links for some reason.)

eddyb replied:
Hang on, `ty::FnPtr` is considered `#[structural_match]`? Are raw pointers as well?

I would understand `ty::FnDef` but this right here is a function *pointer*, isn't it?

cc @pnkfelix

varkor replied:
The [RFC](https://github.com/rust-lang/rfcs/blob/master/text/1445-restrict-constants-in-patterns.md) doesn't even list the types that are intended to be `#[structural_match]`: it just says "builtin types", which would imply function pointers and raw pointers. Certainly that seems to be how `#[structural_match]` is currently implemented: it only really restricts pattern matching on user-defined types.

petrochenkov commented:
This will probably have to wait until the weekend.
r? @petrochenkov

varkor commented on src/librustc/hir/lowering.rs:
```diff
@@ -1149,13 +1149,69 @@ impl<'a> LoweringContext<'a> {
             ast::GenericArg::Lifetime(lt) => GenericArg::Lifetime(self.lower_lifetime(&lt)),
-            ast::GenericArg::Type(ty) => GenericArg::Type(self.lower_ty_direct(&ty, itctx)),
+            ast::GenericArg::Type(ty) => {
+                // A path type argument could be an incorrectly parsed generic const argument. If
```
I think it's better to explain this as: "We parse const arguments as type arguments, because we cannot distinguish between them during parsing. We then try to resolve the ambiguity here." (or similar). Saying it's "incorrectly parsed" makes it sound like the compiler's doing something erroneous.

varkor commented on src/librustc/hir/def.rs:
```diff
@@ -129,6 +129,34 @@ impl DefKind {
         }
     }
+
+    pub fn is_valid_in_ns(&self, ns: Namespace) -> bool {
```
This might be nicer as a `ns` method that returns the namespace (and the checking can be done by the caller).

yodaldevoid replied:
This only problem is with `Res::Err` which is technically valid in all namespaces. I suppose it could return `Option<Namespace` where None is from `Res::Err`.

petrochenkov replied:
Seems good as is, it's used only once and the `bool` semantics matches the use case better.

petrochenkov replied:
`is_valid_in_ns` -> `matches_ns` perhaps?

varkor commented on src/librustc/hir/lowering.rs:
```diff
@@ -1149,13 +1149,69 @@ impl<'a> LoweringContext<'a> {
+                // we resolved the path in the value namespace, we transform it into a generic
+                // const argument.
+                if let TyKind::Path(ref qself, ref path) = ty.kind {
+                    if path.segments.len() == 1 {
```
A comment here explaining this check would be helpful (as consts _can_ have multiple segments, we just can't allow them here).

yodaldevoid replied:
What exactly is the reasoning behind only allowing single segment consts here? I'm not sure I've ever seen the reason spelled out.

petrochenkov replied:
Lowering doesn't care about this condition and can act on the resolution results only (multi-segment paths will never result in values right now, and if they will in the future, they'll need to be treated as consts too), so this line can be removed.

varkor commented on src/librustc_resolve/late.rs:
```diff
@@ -544,6 +544,50 @@ impl<'a, 'tcx> Visitor<'tcx> for LateResolutionVisitor<'a, '_> {
+        match arg {
+            GenericArg::Type(ref ty) => {
+                if let TyKind::Path(ref qself, ref path) = ty.kind {
+                    // A path type argument could be an incorrectly parsed generic const argument.
```
Same comment as above regarding "incorrectly parsed".

varkor commented on src/librustc_resolve/late.rs:
```diff
@@ -544,6 +544,50 @@ impl<'a, 'tcx> Visitor<'tcx> for LateResolutionVisitor<'a, '_> {
+                    // first, and if that fails we try again in the value namespace. If resolution
+                    // in the value namespace succeeds, we have an improperly parsed generic const
+                    // argument on our hands.
+                    if path.segments.len() == 1 {
```
We should duplicate the comment about a single path segment here for clarity.

varkor commented on src/test/ui/const-generics/fn-const-param-infer.stderr:
```diff
@@ -7,10 +7,10 @@ LL | #![feature(const_generics, const_compare_raw_pointers)]
-LL |     let _: Checked<{not_one}> = Checked::<{not_two}>;
-   |                                 ^^^^^^^^^^^^^^^^^^^^ expected `not_one`, found `not_two`
+LL |     let _: Checked<not_one> = Checked::<not_two>;
+   |                               ^^^^^^^^^^^^^^^^^^ expected `not_one`, found `not_two`
```
We have to improve these diagnostics, but that's a separate issue :)

petrochenkov replied:
Not sure what needs to be improved here.

varkor commented on src/test/ui/privacy/privacy-ns1.stderr:
```diff
@@ -86,7 +64,19 @@ LL | use foo2::Bar;
    |
 
-error: aborting due to 4 previous errors
+error[E0107]: wrong number of const arguments: expected 0, found 1
```
This is interesting fallout. I imagine we should be able to special-case this sort of error to get back the old ones, though I'm not sure how much work that would be.

petrochenkov replied:
As far as I remember, the fix may be non-trivial and error-prone, so it's better to make an issue and fix it in a separate PR.

petrochenkov commented on src/librustc/hir/def.rs:
```diff
@@ -129,6 +129,34 @@ impl DefKind {
+            | DefKind::TraitAlias
+            | DefKind::AssocTy
+            | DefKind::AssocOpaqueTy
+            | DefKind::TyParam => Namespace::TypeNS == ns,
```
```suggestion
            | DefKind::TyParam => ns == Namespace::TypeNS,
```
Style nit, here and in few other places.

petrochenkov commented on src/librustc/hir/lowering.rs:
```diff
@@ -1149,13 +1149,69 @@ impl<'a> LoweringContext<'a> {
+                if let TyKind::Path(ref qself, ref path) = ty.kind {
+                    if path.segments.len() == 1 {
+                        if let Some(partial_res) = self.resolver.get_partial_res(ty.id) {
+                            match partial_res.base_res() {
```
The condition should be `!partial_res.base_res().matches_ns(TypeNS)`, then `Res::Err` won't need a special treatment.

I'm pretty sure `CtorKind::Fictive` shouldn't be treated specially too (or should be treated specially in resolve at worst), but I haven't got to the resolve part yet.

yodaldevoid replied:
`!partial_res.base_res().matches_ns(TypeNS)` would also allow it if it had resolved in the macro namespace, correct? Perhaps `!partial_res.base_res().matches_ns(TypeNS) && partial_res.base_res().matches_ns(ValueNS)`?

petrochenkov replied:
All inappropriate resolutions are mapped into `Res::Err` in resolve, so later stages widely assume that you can't get e.g. `Def::Fn` from a type path, we can rely on it here as well.

petrochenkov commented on src/librustc/hir/lowering.rs:
```diff
@@ -1149,13 +1149,69 @@ impl<'a> LoweringContext<'a> {
+                                        DefPathData::AnonConst,
+                                        ExpnId::root(),
+                                        ty.span,
+                                    );
```
Hmm, this distinction between types and consts (consts requiring an extra def path node) is something I didn't expect.
@eddyb, is it possible that creating this node during lowering is too late?

eddyb replied:
Maybe? But if you limit this to a really simple path with no generic args, there shouldn't be any children defs nested inside so making a leaf `AnonConst` should be fine.

petrochenkov replied:
@yodaldevoid 
Could you bring back the "no generic args" restriction then?
(It should rarely be relevant in practice.)

yodaldevoid replied:
It is done.

It should be noted that this does mean that we cannot remove all of the braces from the const-generics tests, specifically in "fn-const-param-infer.rs" which uses type params all over the place. How hard/wise would it be to iterate through the children of the new def to create defs for them if needed?

petrochenkov replied:
>How hard/wise would it be to iterate through the children of the new def to create defs for them if needed?

I'm not sure.
First, we need to at least construct a test where things would break due to def paths not being constructed properly.
This is a material for another PR, IMO.
We get, like, 90% of benefits from disambiguating simple identifiers, so it would be nice to land this PR sooner.

yodaldevoid replied:
I think we already have such a test in "fn-const-param-infer.rs", but I agree that is something for a future PR.

petrochenkov commented on src/librustc/hir/lowering.rs:
```diff
@@ -1149,13 +1149,69 @@ impl<'a> LoweringContext<'a> {
+
+                                    let path_expr = Expr {
+                                        id: ty.id,
+                                        kind: ExprKind::Path(qself.clone(), path.clone()),
```
Can `qself` and `path` be consumed here instead of being cloned?

yodaldevoid replied:
The `GenericArg` we are matching against and getting these parameters from is not consumed, only passed as a reference, so sadly this clone needs to stay.

petrochenkov commented on src/librustc_resolve/late.rs:
```diff
@@ -544,6 +544,50 @@ impl<'a, 'tcx> Visitor<'tcx> for LateResolutionVisitor<'a, '_> {
+                    // in the value namespace succeeds, we have an improperly parsed generic const
+                    // argument on our hands.
+                    if path.segments.len() == 1 {
+                        if let None = self.resolve_qpath(
```
Speculative resolution here should use `resolve_ident_in_lexical_scope` instead of `resolve_qpath`.
See how `PatKind::Ident` are disambiguated using speculative resolution in `resolve_pattern_inner` for an example.
(`resolve_qpath` may leave side effects.)

yodaldevoid replied:
I actually first started with using `resolve_ident_in_lexical_scope`, and swapped over to `resolve_qpath` rather late into these changes. I can fairly easily swap back.

petrochenkov commented on src/librustc_resolve/late.rs:
```diff
@@ -544,6 +544,50 @@ impl<'a, 'tcx> Visitor<'tcx> for LateResolutionVisitor<'a, '_> {
+                            ty.span,
+                            CrateLint::SimplePath(ty.id),
+                        ) {
+                            if let Some(partial_res) = self.resolve_qpath(
```
This should probably be wrapped into `with_constant_rib`.

petrochenkov commented on src/librustc_resolve/late.rs:
```diff
@@ -544,6 +544,50 @@ impl<'a, 'tcx> Visitor<'tcx> for LateResolutionVisitor<'a, '_> {
+                                    "visit_generic_arg: generic type argument resolved as a const"
+                                );
+                                self.r.record_partial_res(ty.id, partial_res);
+                                return;
```
Forgetting to visit generic arguments here.
That may be the reason for the strange behavior in tests where `{}`s could not be removed due to type errors.

petrochenkov replied:
The non-type case should generally mirror treatment of `ExprKind::Path` in `resolve_expr`.

yodaldevoid commented:
I believe I have addressed all comments given so far. There is one [comment](https://github.com/rust-lang/rust/pull/66104#discussion_r344450671) that I would like to have resolved before we move forward with these changes.

petrochenkov commented:
@yodaldevoid 
>What exactly is the reasoning behind only allowing single segment consts here? I'm not sure I've ever seen the reason spelled out.

We need to disambiguate in favor of types, but form multi-segment paths we cannot do that until type checking in general case (due to associated types).
https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/generic.20argument.20disambiguation/near/177847619

We can do some "figure disambiguation" for multi-segment paths to accept *some* cases, but that complicates the rules, and other alternatives like delaying the disambiguation until the generic argument substitution is really required (in type checking) may be preferable.

petrochenkov commented on src/librustc_resolve/late.rs:
```diff
@@ -544,6 +544,60 @@ impl<'a, 'tcx> Visitor<'tcx> for LateResolutionVisitor<'a, '_> {
+                                debug!(
+                                    "visit_generic_arg: generic type argument resolved as a const"
+                                );
+                                self.with_constant_rib(|this| {
```
`with_constant_rib` is supposed to be opened before we resolve the path in value namespace, not after.
This can be tested by attempting to use something non-const, like a local variable, in a generic argument.

petrochenkov replied:
In general, this resolution in value namespace should reuse code from `visit_anon_const` as much as possible instead of adding its own subtly different code path.
This means `with_constant_rib` on the outside, then copypaste of the `ExprKind::Path` case from `resolve_expr`.
(Unfortunately, we cannot use `visit_anon_const` directly here due to visitor lifetimes.)

yodaldevoid replied:
I have attempted to wrap the `with_constant_rib` around the resolution in the value namespace, but have found it not to work quite right.

If I use `smart_resolve_path` as done in the `ExprKind::Path` branch of `resolve_expr` for the value namespace resolution, core doesn't even compile. This is due to macros breaking as all generic args in macros start to be seen as values, even where they should be types. I had avoided `smart_resolve_path` to begin with because that smartness includes throwing a bunch of errors that would be erroneous in this case and it doesn't provide any feedback about success to the caller.

If I use `resolve_ident_in_lexical_scope` for the value namespace resolution everything compiles fine, except the macros in the test "ui/macros/type-macros-simple.rs". In this case the constant rib starts complaining that the primitive types are not constant values.

Now, this is all with using `resolve_ident_in_lexical_scope` for type value resolution. If `resolve_qpath` is used instead, as it is done currently in the patches, everything is fine when using `smart_resolve_path` for value namespace resolution, though there are quite a few error message changes seen in tests. I have pushed these changes up for everyone to review.

petrochenkov replied:
Ok, I'll look what happens when `resolve_ident_in_lexical_scope` is used for the speculative check in type namespace, because what you describe sounds strange.

>when using smart_resolve_path for value namespace resolution, though there are quite a few error message changes seen in tests.

Yeah, we'll probably have to use two speculative checks to improve the diagnostics for the case when the name cannot be found in both namespaces.
Basically, side-effectful `smart_resolve_path` for values should not be called if a "pure" speculative check fails in value namespace. Then we'll still get "unresolved type" errors and not "unresolved value".

petrochenkov replied:
Here's the fix - https://github.com/petrochenkov/rust/commit/7c3a4027e9f527151297e6c70b54e67885ebce47.
`resolve_ident_in_lexical_scope` had an issue with built-in types that was generally masked in `resolve_qpath` a couple of levels above, but still prevented speculative use of `resolve_ident_in_lexical_scope` in type namespace.

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=13327&view=logs&jobId=d0d954b5-f111-5dc4-4d76-03b6c9d0cf7e) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/13327/logs/67), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/13327/logs/67)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2019-11-12T02:16:46.3004959Z ##[command]git remote add origin https://github.com/rust-lang/rust
2019-11-12T02:16:46.3210629Z ##[command]git config gc.auto 0
2019-11-12T02:16:46.3293207Z ##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
2019-11-12T02:16:46.3352264Z ##[command]git config --get-all http.proxy
2019-11-12T02:16:47.2218375Z ##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/66104/merge:refs/remotes/pull/66104/merge
---
2019-11-12T02:27:23.0096163Z      |
2019-11-12T02:27:23.0096664Z 1183 | ...                   let node_id = self.sess.next_node_id();
2019-11-12T02:27:23.0097506Z      |                                               ^^^^^^^^^^^^ method not found in `&'a session::Session`
2019-11-12T02:27:23.0097977Z      |
2019-11-12T02:27:23.0098464Z      = help: items from traits can only be used if the trait is implemented and in scope
2019-11-12T02:27:23.0098939Z      = note: the following trait defines an item `next_node_id`, perhaps you need to implement it:
2019-11-12T02:27:23.0099367Z              candidate #1: `hir::lowering::Resolver`
2019-11-12T02:27:35.7738498Z error: aborting due to previous error
2019-11-12T02:27:35.7740732Z 
2019-11-12T02:27:35.7741560Z For more information about this error, try `rustc --explain E0599`.
2019-11-12T02:27:35.9378778Z error: could not compile `rustc`.
---
2019-11-12T02:27:41.6686661Z   local time: Tue Nov 12 02:27:41 UTC 2019
2019-11-12T02:27:41.6972523Z   network time: Tue, 12 Nov 2019 02:27:41 GMT
2019-11-12T02:27:41.6978563Z == end clock drift check ==
2019-11-12T02:27:42.6574530Z 
2019-11-12T02:27:42.6713638Z ##[error]Bash exited with code '1'.
2019-11-12T02:27:42.6738566Z ##[section]Starting: Checkout
2019-11-12T02:27:42.6740515Z ==============================================================================
2019-11-12T02:27:42.6740571Z Task         : Get sources
2019-11-12T02:27:42.6740638Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

yodaldevoid commented on src/librustc_resolve/late.rs:
```diff
@@ -544,6 +544,60 @@ impl<'a, 'tcx> Visitor<'tcx> for LateResolutionVisitor<'a, '_> {
+                    // Const arguments cannot have a multi-segment path right now, so we check that
+                    // first.
+                    if path.segments.len() == 1 {
+                        if let None = self.resolve_qpath(
```
See https://github.com/rust-lang/rust/pull/66104#discussion_r344998068

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=13330&view=logs&jobId=d0d954b5-f111-5dc4-4d76-03b6c9d0cf7e) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/13330/logs/67), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/13330/logs/67)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2019-11-12T02:55:51.5195173Z ##[command]git remote add origin https://github.com/rust-lang/rust
2019-11-12T02:55:51.5403722Z ##[command]git config gc.auto 0
2019-11-12T02:55:51.5475762Z ##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
2019-11-12T02:55:51.5549834Z ##[command]git config --get-all http.proxy
2019-11-12T02:55:51.5683693Z ##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/66104/merge:refs/remotes/pull/66104/merge
---
2019-11-12T03:06:29.3048553Z      |
2019-11-12T03:06:29.3048918Z 1183 | ...                   let node_id = self.sess.next_node_id();
2019-11-12T03:06:29.3049333Z      |                                               ^^^^^^^^^^^^ method not found in `&'a session::Session`
2019-11-12T03:06:29.3049587Z      |
2019-11-12T03:06:29.3050073Z      = help: items from traits can only be used if the trait is implemented and in scope
2019-11-12T03:06:29.3050383Z      = note: the following trait defines an item `next_node_id`, perhaps you need to implement it:
2019-11-12T03:06:29.3050610Z              candidate #1: `hir::lowering::Resolver`
2019-11-12T03:06:41.7255823Z error: aborting due to previous error
2019-11-12T03:06:41.7257403Z 
2019-11-12T03:06:41.7258504Z For more information about this error, try `rustc --explain E0599`.
2019-11-12T03:06:41.8788653Z error: could not compile `rustc`.
---
2019-11-12T03:06:47.0321850Z   local time: Tue Nov 12 03:06:47 UTC 2019
2019-11-12T03:06:47.1185108Z   network time: Tue, 12 Nov 2019 03:06:47 GMT
2019-11-12T03:06:47.1185229Z == end clock drift check ==
2019-11-12T03:06:48.1766653Z 
2019-11-12T03:06:48.1880383Z ##[error]Bash exited with code '1'.
2019-11-12T03:06:48.1908605Z ##[section]Starting: Checkout
2019-11-12T03:06:48.1910277Z ==============================================================================
2019-11-12T03:06:48.1910322Z Task         : Get sources
2019-11-12T03:06:48.1910361Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=13338&view=logs&jobId=d0d954b5-f111-5dc4-4d76-03b6c9d0cf7e) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/13338/logs/67), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/13338/logs/67)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2019-11-12T05:48:43.8941057Z ##[command]git remote add origin https://github.com/rust-lang/rust
2019-11-12T05:48:44.4392170Z ##[command]git config gc.auto 0
2019-11-12T05:48:44.4396747Z ##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
2019-11-12T05:48:44.4398974Z ##[command]git config --get-all http.proxy
2019-11-12T05:48:44.4404848Z ##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/66104/merge:refs/remotes/pull/66104/merge
---
2019-11-12T05:59:05.1218613Z      |
2019-11-12T05:59:05.1218947Z 1183 | ...                   let node_id = self.sess.next_node_id();
2019-11-12T05:59:05.1219376Z      |                                               ^^^^^^^^^^^^ method not found in `&'a session::Session`
2019-11-12T05:59:05.1219611Z      |
2019-11-12T05:59:05.1219975Z      = help: items from traits can only be used if the trait is implemented and in scope
2019-11-12T05:59:05.1220309Z      = note: the following trait defines an item `next_node_id`, perhaps you need to implement it:
2019-11-12T05:59:05.1220571Z              candidate #1: `hir::lowering::Resolver`
2019-11-12T05:59:17.6554572Z error: aborting due to previous error
2019-11-12T05:59:17.6559101Z 
2019-11-12T05:59:17.6565596Z For more information about this error, try `rustc --explain E0599`.
2019-11-12T05:59:17.8230361Z error: could not compile `rustc`.
---
2019-11-12T05:59:22.7322623Z   local time: Tue Nov 12 05:59:22 UTC 2019
2019-11-12T05:59:22.8053015Z   network time: Tue, 12 Nov 2019 05:59:22 GMT
2019-11-12T05:59:22.8053121Z == end clock drift check ==
2019-11-12T05:59:23.8633026Z 
2019-11-12T05:59:23.8738770Z ##[error]Bash exited with code '1'.
2019-11-12T05:59:23.8770389Z ##[section]Starting: Checkout
2019-11-12T05:59:23.8772611Z ==============================================================================
2019-11-12T05:59:23.8772695Z Task         : Get sources
2019-11-12T05:59:23.8772748Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

yodaldevoid commented:
In regards to the current build failures, I have no idea what is going on. It builds for me on both a Windows and Linux machine, one of which is a fresh checkout. I might try rebasing later to see if that fixes anything.

petrochenkov commented:
Could you squash commits after addressing the remaining comments?

petrochenkov commented:
@bors r+

bors commented:
:pushpin: Commit 0207a15fa14c2c05e33acac1abd4604fce1f346a has been approved by `petrochenkov`

<!-- @bors r=petrochenkov 0207a15fa14c2c05e33acac1abd4604fce1f346a -->
<!-- homu: {"type":"Approved","sha":"0207a15fa14c2c05e33acac1abd4604fce1f346a","approver":"petrochenkov"} -->

bors commented:
:hourglass: Testing commit 0207a15fa14c2c05e33acac1abd4604fce1f346a with merge 0a431b8e96bfc211dd2247e3cd56105e589b13a2...
<!-- homu: {"type":"BuildStarted","head_sha":"0207a15fa14c2c05e33acac1abd4604fce1f346a","merge_sha":"0a431b8e96bfc211dd2247e3cd56105e589b13a2"} -->

rust-highfive commented:
Your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=13920&view=logs&jobId=a1bbfa4f-3958-53e8-464f-18f9feaa3431) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/13920/logs/1107), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/13920/logs/1107)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2019-11-19T18:41:44.6383257Z Attempting with retry: docker build --rm -t rust-ci -f /home/vsts/work/1/s/src/ci/docker/dist-x86_64-netbsd/Dockerfile /home/vsts/work/1/s/src/ci/docker
2019-11-19T18:41:44.8247314Z Sending build context to Docker daemon  528.4kB
2019-11-19T18:41:44.8247599Z 
2019-11-19T18:41:44.8847651Z Step 1/12 : FROM ubuntu:16.04
2019-11-19T18:41:45.5615847Z Get https://registry-1.docker.io/v2/library/ubuntu/manifests/16.04: received unexpected HTTP status: 500 Internal Server Error
2019-11-19T18:41:46.7933149Z Sending build context to Docker daemon  528.4kB
2019-11-19T18:41:46.7933379Z 
2019-11-19T18:41:46.8166438Z Step 1/12 : FROM ubuntu:16.04
2019-11-19T18:41:46.8166438Z Step 1/12 : FROM ubuntu:16.04
2019-11-19T18:41:47.3936219Z Get https://registry-1.docker.io/v2/library/ubuntu/manifests/16.04: received unexpected HTTP status: 500 Internal Server Error
2019-11-19T18:41:49.5810512Z Sending build context to Docker daemon  528.4kB
2019-11-19T18:41:49.5810778Z 
2019-11-19T18:41:49.6004519Z Step 1/12 : FROM ubuntu:16.04
2019-11-19T18:41:49.6004519Z Step 1/12 : FROM ubuntu:16.04
2019-11-19T18:41:50.0978698Z Get https://registry-1.docker.io/v2/library/ubuntu/manifests/16.04: received unexpected HTTP status: 500 Internal Server Error
2019-11-19T18:41:53.2482214Z Sending build context to Docker daemon  528.4kB
2019-11-19T18:41:53.2487663Z 
2019-11-19T18:41:53.2766217Z Step 1/12 : FROM ubuntu:16.04
2019-11-19T18:41:53.2766217Z Step 1/12 : FROM ubuntu:16.04
2019-11-19T18:41:53.7856733Z Get https://registry-1.docker.io/v2/library/ubuntu/manifests/16.04: received unexpected HTTP status: 500 Internal Server Error
2019-11-19T18:41:57.9035917Z Sending build context to Docker daemon  528.4kB
2019-11-19T18:41:57.9036879Z 
2019-11-19T18:41:57.9245669Z Step 1/12 : FROM ubuntu:16.04
2019-11-19T18:41:57.9245669Z Step 1/12 : FROM ubuntu:16.04
2019-11-19T18:41:58.4398790Z Get https://registry-1.docker.io/v2/library/ubuntu/manifests/16.04: received unexpected HTTP status: 500 Internal Server Error
2019-11-19T18:41:58.4470573Z 
2019-11-19T18:41:58.4470573Z 
2019-11-19T18:41:58.4581131Z ##[error]Bash exited with code '1'.
2019-11-19T18:41:58.4614098Z ##[section]Starting: Checkout
2019-11-19T18:41:58.4616398Z ==============================================================================
2019-11-19T18:41:58.4616517Z Task         : Get sources
2019-11-19T18:41:58.4616601Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

bors commented:
:broken_heart: Test failed - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=13920)
<!-- homu: {"type":"BuildFailed","builder_url":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=13920","builder_name":"checks-azure"} -->

varkor commented:
@bors retry

bors commented:
:hourglass: Testing commit 0207a15fa14c2c05e33acac1abd4604fce1f346a with merge f50d6ea348c2dd7c2f76e35ecde6560d87bb98ec...
<!-- homu: {"type":"BuildStarted","head_sha":"0207a15fa14c2c05e33acac1abd4604fce1f346a","merge_sha":"f50d6ea348c2dd7c2f76e35ecde6560d87bb98ec"} -->

bors commented:
:sunny: Test successful - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=13947)
Approved by: petrochenkov
Pushing f50d6ea348c2dd7c2f76e35ecde6560d87bb98ec to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"petrochenkov","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=13947"},"merge_sha":"f50d6ea348c2dd7c2f76e35ecde6560d87bb98ec"} -->

rust-highfive commented:
📣 Toolstate changed by rust-lang/rust#66104!

Tested on commit rust-lang/rust@f50d6ea348c2dd7c2f76e35ecde6560d87bb98ec.
Direct link to PR: <https://github.com/rust-lang/rust/pull/66104>

🎉 rustc-guide on linux: test-fail → test-pass (cc @JohnTitor @amanjeev @spastorino @mark-i-m, @rust-lang/infra).
